model/DeliveryContainerService.php

```diff
@@ -88,30 +93,33 @@ function ($provider) {
      * Get the list of active plugins for the current execution
      * @param DeliveryExecution $deliveryExecution
      * @return TestPlugin[] the list of plugins
-     */
-    /**
-     * @param DeliveryExecution $deliveryExecution
-     * @return array
+     *
      * @throws \common_exception_NotFound
      */
     public function getPlugins(DeliveryExecution $deliveryExecution)
     {
-        $pluginService = $this->getServiceLocator()->get(TestPluginService::SERVICE_ID);
-        $allPlugins = $pluginService->getAllPlugins();
-
-        $pluginsToDisable = $this->getPluginsDisabledForDelivery($deliveryExecution->getDelivery());
-        if (!empty($pluginsToDisable)) {
-            foreach ($allPlugins as $plugin) {
-                if (!is_null($plugin) && in_array($plugin->getId(), $pluginsToDisable)) {
-                    $plugin->setActive(false);
+        $delivery = $deliveryExecution->getDelivery();
+        $deliveryUri = $delivery->getUri();
+        if (!isset($this->deliveryPlugins[$deliveryUri])) {
+            $pluginService = $this->getServiceLocator()->get(TestPluginService::SERVICE_ID);
+            $allPlugins = $pluginService->getAllPlugins();
+
+            $pluginsToDisable = $this->getPluginsDisabledForDelivery($delivery);
+            if (!empty($pluginsToDisable)) {
+                foreach ($allPlugins as $plugin) {
+                    if (!is_null($plugin) && in_array($plugin->getId(), $pluginsToDisable)) {
+                        $plugin->setActive(false);
+                    }
                 }
             }
+
+            // return only list of active plugins
+            $this->deliveryPlugins[$deliveryUri] = array_filter($allPlugins, function ($plugin) {
+                return !is_null($plugin) && $plugin->isActive();
+            });
```

Comment:
Maybe we can do everything during one `foreach` instead of `foreach` + `array_filter`?

```
$pluginsToDisable = $this->getPluginsDisabledForDelivery($delivery);

foreach ($allPlugins as $plugin) {
    if ($plugin !== null) {
        if (in_array($plugin->getId(), $pluginsToDisable, true)) {
            $plugin->setActive(false);
        } elseif ($plugin->isActive()) {
            $this->deliveryPlugins[$deliveryUri][] = $plugin;
        }
    }
}
```

Reply:
done